File: c/data-structures/7_queue.cpp
#include "utils.h"

struct Element {
  int data;
  Element* next;
};

struct Queue {
  size_t len = 0;
  size_t cap = 0;
  Element* front = nullptr;
};

bool is_clear(Queue &queue) {
  return queue.len == 0;
}

bool is_full(Queue &queue) {
  return queue.len == queue.cap;
}

Element* create(int data, size_t index) {
  Element* element = new Element;
  element->data = data;
  element->next = nullptr;
  return element;
}

Queue create_queue(size_t cap) {
  Queue queue;
  queue.cap = cap;
  return queue;
}

void enqueue(Queue &queue, int data) {
  if (is_full(queue)) {
    throw std::overflow_error("Cannot enqueue: Queue is full");
  }

  Element* element = create(data, queue.len);

  if (is_clear(queue)) {
    queue.front = element;
  } else {
    element->next = queue.front;
    queue.front = element;
  }

  queue.len++;
}

Element* dequeue(Queue &queue) {
  if (is_clear(queue)) {
    throw std::underflow_error("Cannot dequeue: Queue is empty");
  }

  Element *front = queue.front;
  queue.front = front->next;
  queue.len++;
  front->next = nullptr;

  return front;
}

void clear(Queue &queue) {
  Element* current = queue.front;
  
  while (current != nullptr) {
    Element* next = current->next;
    delete current;
    current = next;
  }
}

void print_element(Element* element, size_t indentation = 0) {
  if (element == nullptr) return;

  std::string indent = std::string(indentation, ' ');

  if (element->next == nullptr) {
    printf("Element { data: %d }\n", element->data);
  } else {
    println("Element {");
    println(indent + "  data: " + std::to_string(element->data));
    print(indent + "  next: ");
    print_element(element->next, indentation + 2);
    println(indent + "}");
  }
}

void print_queue(Queue queue) {
  print_element(queue.front);
}

int main() {
  size_t cap = get_input<size_t>("Enter the capacity of the queue");
  Queue queue = create_queue(cap);

  while (true) {
    println("What do you want to do?");
    println("1. Enqueue");
    println("2. Dequeue");
    println("3. Print");
    println("4. Exit");

    int option = get_input<int>(">> Enter an option");

    switch (option) {
      case 1: {
        int data = get_input<int>("Enter the data");
        enqueue(queue, data);
        break;
      }
      case 2: {
        Element* element = dequeue(queue);
        println("Dequed Element:");
        print_element(element);
        delete element;
        break;
      }
      case 3: {
        print_queue(queue);
        break;
      }
      case 4: 
        clear(queue);
        println("Bye!");
        return 0;
      default: {
        println("Invalid option");
        break;
      }
    }

  };

  return 0;
}

File: python/data-structures/7_queue.py
from utils import get_str, get_int

class Element:
    data: int
    next: "Element"

    def __init__(self, data: int):
        self.data = data
        self.next = None

    def print(self, indentation = 0):
        indent = " " * indentation

        if self.next == None:
            print(f"Element {{ data: {self.data} }}")
        else:
            print("Element {")
            print(indent + f"    data: {self.data}")
            print(indent + "    next: ", end = "")
            self.next.print(indentation + 2)
            print(indent + "  }")


class Queue:
    cap: int
    len: int
    front: Element

    def __init__(self, cap: int):
        self.cap = cap
        self.len = 0

    def enqueue(self, data: int):
        if self.len == self.cap:
            raise "Cannot enqueue element: Queue is full"

        child = Element(data)

        if self.len == 0:
            self.front = child
        else:
            child.next = self.front
            self.front = child

        self.len += 1

    def dequeue(self) -> Element:
        if self.len == 0:
            return None

        front = self.front
        self.front = self.front.next
        front.next = None
        self.len -= 1

        return front

    def has_value(self, data: int):
        current = self.front

        while current != None:
            if current.data == data:
                return True

            current = current.next

        return False

    def clear(self):
        self.front = None
        self.len = 0

    def print(self):
        print("Queue {")
        print(f"  cap: {self.cap}")
        print(f"  len: {self.len}")

        if self.front:
            print("  front: ", end = "")
            self.front.print()

        print("}")


cap = get_int("Enter the capacity of the queue")
queue = Queue(cap)

while True:
    print("What do you want to do?")
    print("1. Enqueue")
    print("2. Dequeue")
    print("3. Print")
    print("4. Exit")

    option = get_int(">>> Enter the option")

    if option == 1:
        data = get_int("Enter the data")
        queue.enqueue(data)
        continue

    if option == 2:
        element = queue.dequeue()
        print("Dequeued Element")
        element.print()
        continue

    if option == 3:
        queue.print()
        continue

    if option == 4:
        queue.clear()
        print("Bye")
        break

    print("Invalid Option!")
